lib/semantics/tools.h

```diff
@@ -113,10 +113,49 @@ inline bool IsAssumedSizeArray(const Symbol &symbol) {
 }
 bool IsAssumedLengthCharacter(const Symbol &);
 bool IsAssumedLengthCharacterFunction(const Symbol &);
+// Is the symbol modifiable in this scope
 std::optional<parser::MessageFixedText> WhyNotModifiable(
     const Symbol &symbol, const Scope &scope);
-// Is the symbol modifiable in this scope
 bool IsExternalInPureContext(const Symbol &symbol, const Scope &scope);
+bool HasCoarray(const parser::Expr &expression);
+template<typename T>
+
+// Detection of image control statements.  If the parameter is not an image
+// control statement, you get nullopt.  Otherwise, you get a message that
+// gives the specifics of why the construct is not allowed.
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(const T &) {
+  return std::nullopt;
+}
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
+    const parser::SyncAllStmt &);
+std::optional<parser::MessageFixedText> GetImageControlStmtMsg(
```

Comment:
Roughly a dozen of these simply return the same error string with the construct name pasted on the front. You could put these all together in a template function that returns a generic "not allowed" message.

That is complicated by the calling of `GetImageControlStmtMsg` on anything, rather than selectively on something known to be an image control statement.  Separation of concerns here would be to have one method determine *if* the object is an offender and a distinct method that constructs a suitable error message/report.

Reply:
Thanks, Eric.

This is similar to Peter's comment above.  I'm currently working on making this change.